### ISSUE TITLE
MOD operator returns wrong results for negative argument

### ISSUE THREAD

**Issue by tingtron** [robhagemans/pcbasic#231]
## Bug report

**Problem**
MOD operator returns wrong results with negative argument.

**Steps**
1. Run PC-BASIC  
2.  Enter the following and press `Enter`
 `for i=-5 to 3: ? i mod 2; : next`
3. Observed: `-1 -2 -1 -2 -1  0  1  0  1` 
4. Expected:  `-1 0 -1 0 -1  0  1  0  1`
5. Compare with GW-BASIC from MS-DOS (in DOSBox)

Note: Python works as expected:
```
>>> [x % 2 for x in range(-5,3)]
[1, 0, 1, 0, 1, 0, 1, 0]
```

The problem is in the source `pcbasic\basic\values\numbers.py`:
>         if dividend < 0 or mod < 0:
>           mod -= divisor
It should be
>           mod = -mod

**Program**
This affects an existing program, which calculates coordinate offset.

**Notes**
PC-BASIC version:  2.0.7
Operating system version:  Windows 11 64-bit

**Comment by robhagemans** [comment]
I can confirm this is a bug and I can reproduce.

**Comment by robhagemans** [comment]
For reference, these are the results for Python and for GW-BASIC:

GW-BASIC:
```
? 17 mod 3                                                                      
 2                                                                              
Ok                                                                              
? 17 mod -3                                                                     
 2                                                                              
Ok                                                                              
? -17 mod 3                                                                     
-2                                                                              
Ok                                                                              
? -17 mod -3                                                                    
-2                                                                              
Ok     
```
Python:
```
>>> 17 % 3
2
>>> 17 % -3
-1
>>> -17 % 3
1
>>> -17 % -3
-2
```

Some background as to what Python does and why it does that, linked below. TLDR, Python's `//` rounds toward negative infinity, and `%` then does what it needs to do to ensure the relation holds that `divisor*quotient + remainder = dividend`.
https://stackoverflow.com/questions/3883004/how-does-the-modulo-operator-work-on-negative-numbers-in-python#3883019
http://python-history.blogspot.com/2010/08/why-pythons-integer-division-floors.html

**Comment by tingtron** [comment]
Thank you for looking into this.

For QBasic purposes, it would make sense to replicate the original behavior,
as opposed to trying to find some mathematical truth. The reason being that
a game written in QBasic would be expected to run identically on PC-BASIC.

Your reference from QBasic make sense.

Notably, sign of the divisor does not play a role, only the sign of dividend.

As shown with the output of the following example:
``` basic
-3 mod 3 = 0 
-2 mod 3 =-2 
-1 mod 3 =-1 
 0 mod 3 = 0 
 1 mod 3 = 1 
 2 mod 3 = 2 
 3 mod 3 = 0 

-3 mod -3 = 0 
-2 mod -3 =-2 
-1 mod -3 =-1 
 0 mod -3 = 0 
 1 mod -3 = 1 
 2 mod -3 = 2 
 3 mod -3 = 0 
```

For reference, the command line version of the QBasic example above.
- negmod.bas
``` basic
FOR i = -3 TO 3: PRINT i; "mod 3 ="; i MOD 3: NEXT: PRINT
FOR i = -3 TO 3: PRINT i; "mod -3 ="; i MOD -3: NEXT: PRINT

SYSTEM
```

And its command line:
``` cmd
qbasic /run negmod > negmod.txt 
```

**Comment by robhagemans** [comment]
Yes, that is the spirit of the fix that has been merged. Note that PC-BASIC aims to replicate GW-BASIC, not QBASIC, although I expect the behave the same as regards the MOD operator.
With the fix, the behaviour of PC-BASIC matches GW-BASIC.

**Comment by tingtron** [comment]
Right, GW-BASIC.

Corresponding test script:
``` basic
10 FOR i = -3 TO 3: PRINT i; "mod 3 ="; i MOD 3: NEXT: PRINT
20 FOR i = -3 TO 3: PRINT i; "mod -3 ="; i MOD -3: NEXT: PRINT
30 SYSTEM
```
run as
``` cmd
gwbasic negmod.bas > negmod.txt
```
produces the same output:
```
-3 mod 3 = 0 
-2 mod 3 =-2 
-1 mod 3 =-1 
 0 mod 3 = 0 
 1 mod 3 = 1 
 2 mod 3 = 2 
 3 mod 3 = 0 

-3 mod -3 = 0 
-2 mod -3 =-2 
-1 mod -3 =-1 
 0 mod -3 = 0 
 1 mod -3 = 1 
 2 mod -3 = 2 
 3 mod -3 = 0 
```

**Comment by robhagemans** [comment]
Yes, as I said. This has been implemented. what is your point please?

**Comment by tingtron** [comment]
It is for clarification purposes here. I was reading the _original_ issue, and noticed a few omissions: it did not differentiate between positive and negative divisors, hence the suggested formula was also wrong. The test cases in the committed sources reflect the correct approach.

Not sure about the process to close the issue -- if you verified it, or? But will probably we waiting for the next release to verify. Unless there is a 'beta' release that I could test.

**Comment by robhagemans** [comment]
I'm keeping it open for now, with the "fixed" label, so people can see it's a known issue and don't report it again. If/when I do another release I'll close. meanwhile, if you do want to test it, it's on the master branch in the repository here on github

**Comment by robhagemans** [comment]
Fix included in v2.0.8
